contentcuration/contentcuration/static/js/exercise_creation/views.js

```diff
@@ -0,0 +1,562 @@
+        }
+
+        // TODO parameterize to allow different file uploads depending on initialization.
+        this.dropzone = new Dropzone(this.$("#dropzone").get(0), {maxFiles: 1, clickable: ["#dropzone", ".fileinput-button"], acceptedFiles: "image/*", url: window.Urls.file_upload(), headers: {"X-CSRFToken": get_cookie("csrftoken")}});
+        this.dropzone.on("success", this.file_uploaded);
+
+    },
+
+    file_uploaded: function(file) {
+        this.callback(JSON.parse(file.xhr.response).filename);
+        this.close();
+    },
+
+    close: function() {
+        if (this.modal) {
+            this.$(".modal").modal('hide');
+        }
+        this.remove();
+    }
+});
+
+var EditorView = Backbone.View.extend({
+
+    tagName: "div",
+
+    initialize: function(options) {
+        _.bindAll(this, "return_markdown", "add_image", "deactivate_editor", "activate_editor", "save_and_close", "save", "render");
+        this.edit_key = options.edit_key;
+        this.editing = false;
+        this.render();
+        this.listenTo(this.model, "change:" + this.edit_key, this.render);
+    },
+
+    events: {
+        "click .ql-image": "add_image_popup"
+    },
+
+    add_image_popup: function() {
+        var view = new FileUploadView({callback: this.add_image, modal: true});
+    },
+
+    add_image: function(filename) {
+        this.editor.insertEmbed(this.editor.getSelection() !== null ? this.editor.getSelection().start : this.editor.getLength(), "image", "/media/" + filename);
+        this.save();
+    },
+
+    edit_template: require("./hbtemplates/editor.handlebars"),
+
+    view_template: require("./hbtemplates/editor_view.handlebars"),
+
+    render: function() {
+        if (this.editing) {
+            if (!this.setting_model) {
+                this.render_editor();
+            }
+        } else {
+            this.render_content();
+        }
+        this.setting_model = false;
+    },
+
+    render_content: function() {
+        this.$el.html(this.view_template({content: this.model.get(this.edit_key)}));
+    },
+
+    render_editor: function() {
+        this.editor.setHTML(this.view_template({content: this.model.get(this.edit_key)}));
+    },
+
+    activate_editor: function() {
+        this.$el.html(this.edit_template());
+        this.editor = new Quill(this.$(".editor")[0], {
+            modules: {
+                'toolbar': { container: this.$('#toolbar')[0] }
+            },
+            theme: 'snow'
+        });
+        this.render_editor();
+        this.editor.on("text-change", _.debounce(this.save, 500));
+        this.editing = true;
+    },
+
+    deactivate_editor: function() {
+        delete this.editor;
+        this.editing = false;
+        this.render();
+    },
+
+    toggle_editor: function() {
+        if (this.editor) {
+            this.deactivate_editor();
+        } else {
+            this.activate_editor();
+        }
+    },
+
+    save: function(delta, source) {
```

Comment:
These are the arguments passed by QuillJS (the editor library I am using) - delta is the change that has occurred and source is the source of that change. This is just to make sure we don't end up in render driven feedback loops.